### PR TITLE
[Memory Planner] Update the Memory Planner 

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes,
  1. assigning right execution order depending on layer type.
     : We do need to move those fixes in to each layer
  2. Update memory Planner to use the memory with smaller size

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Review by baek2sm — APPROVED** [review]
LGTM

File: nntrainer/graph/graph_node.h [hunk 1]
```diff
@@ -39,7 +39,8 @@ class GraphNode {
    * This ensures that the operations are executed in the order of their
    * listing.
    */
-  typedef std::tuple<unsigned int, unsigned int, unsigned int, unsigned int> ExecutionOrder;
+  typedef std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>
```
Comment (SeoHyungjun) [inline nntrainer/graph/graph_node.h:42]: Wouldn't it be more concise if changed unsigned int to uint?

**Review by SeoHyungjun — APPROVED** [review]
LGTM!

File: nntrainer/graph/graph_node.h (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/graph/graph_node.h:42]: we are using unsigned int so many point in NNTrainer. may be separate PR might better to change everywhere.

File: nntrainer/tensor/optimized_v1_planner.cpp [hunk 2]
```diff
@@ -192,6 +215,65 @@ size_t OptimizedV1Planner::planLayout(
...
+        if (merge) {
+          req.offset = sr.mem_req->offset;
+          memory_offset[req.loc] = req.offset;
+          replace_and_fill = true;
+          wgrad_sorted_req[idx].start_end.push_back(
+            std::make_pair(req.start, req.end));
+          break;
+        } else {
+          replace_and_fill = false;
+        }
```
Comment (jihochu) [inline nntrainer/tensor/optimized_v1_planner.cpp]: else loop is unnecessary.

File: nntrainer/tensor/manager.cpp [hunk 3]
```diff
@@ -534,9 +541,18 @@ Manager::requestInputs(const GraphNode &node,
...
       node.getType() == MultiOutLayer::type or
       node.getType() == BatchNormalizationLayer::type or
-      node.getType() == LayerNormalizationLayer::type)
+      node.getType() == LayerNormalizationLayer::type or !node.getTrainable())
     var_common_spec.ls = TensorLifespan::FORWARD_FUNC_LIFESPAN;
 
+  if (node.getType() == MSELossLayer::type or
+      node.getType() == CrossEntropySoftmaxLossLayer::type or
+      node.getType() == CrossEntropySigmoidLossLayer::type)
+    var_common_spec.ls = TensorLifespan::FORWARD_DERIV_LIFESPAN;
```
Comment (jihochu) [inline nntrainer/tensor/manager.cpp:550]: Looks good.

File: nntrainer/tensor/optimized_v2_planner.cpp [hunk 4]
```diff
@@ -282,15 +286,11 @@ size_t OptimizedV2Planner::planLayout(
       wgrad_sorted_req.push_back(WGradMemoryRequest(&req));
       wgrad_sorted_req.back().start_end.push_back(
         std::make_pair(req.start, req.end));
+#ifdef DEBUG
       new_grad_cnt++;
       new_grad_size += req.size;
+#endif
     }
-
-    ml_logd("Total Requested Memory(OPTV2): %lf MiB>>>>>>>> \n - new mem for "
```
Comment (jihochu) [inline nntrainer/tensor/optimized_v2_planner.cpp]: Isn't it better to keep this log for debugging?

File: nntrainer/layers/layer_context.h [hunk 5]
```diff
@@ -235,7 +235,7 @@ class InitLayerContext {
   static VarGradSpecV2
   outSpec(const TensorDim &dim, const std::string &name = "out",
           TensorLifespan ls = TensorLifespan::FORWARD_FUNC_LIFESPAN,
-          TensorLifespan grad_ls = TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+          TensorLifespan grad_ls = TensorLifespan::CALC_GRAD_DERIV_LIFESPAN);
```
Comment (jihochu) [inline nntrainer/layers/layer_context.h:238]: `BACKWARD_FUNC_LIFESPAN` contains `apply gradient` step.
IMHO, Isn't it better to remain it as default while `calc grad` is set?

Comment (jijoongmoon) [reply nntrainer/layers/layer_context.h:238]: In that case, we will have unnecessary execution order for the input tensor, which is not desirable.  We do not need the outgoing gradient to compute the apply gradient.

File: nntrainer/tensor/optimized_v2_planner.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/tensor/optimized_v2_planner.cpp]: well, may be, it's for later pr, then.

File: nntrainer/tensor/optimized_v1_planner.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/tensor/optimized_v1_planner.cpp]: removed. Thanks.

File: nntrainer/layers/layer_context.h (hunk 5 above)
Comment (jihochu) [reply nntrainer/layers/layer_context.h:238]: OK. thanks fo checking.

**Review by jihochu — APPROVED** [review]
LGTM

**Pull request merged**